onmetal-controller-manager/controllers/networking/natgateway_controller.go

```diff
@@ -0,0 +1,318 @@
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	MIN_PORT int32 = 1024
+	MAX_PORT int32 = 65535
+)
+
+var (
+	natGatewayFieldOwner = client.FieldOwner(networkingv1alpha1.Resource("natgateways").String())
+)
+
+type NatGatewayReconciler struct {
+	record.EventRecorder
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=networking.api.onmetal.de,resources=natgateways,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=networking.api.onmetal.de,resources=natgateways/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=networking.api.onmetal.de,resources=natgateways/finalizers,verbs=update
+//+kubebuilder:rbac:groups=networking.api.onmetal.de,resources=networkinterfaces,verbs=get;list;watch
+//+kubebuilder:rbac:groups=networking.api.onmetal.de,resources=natgatewayroutings,verbs=get;list;watch;create;update;patch;delete
+
+func (r *NatGatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	natGateway := &networkingv1alpha1.NATGateway{}
+	if err := r.Get(ctx, req.NamespacedName, natGateway); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	return r.reconcileExists(ctx, log, natGateway)
+}
+
+func (r *NatGatewayReconciler) reconcileExists(ctx context.Context, log logr.Logger, natGateway *networkingv1alpha1.NATGateway) (ctrl.Result, error) {
+	if !natGateway.DeletionTimestamp.IsZero() {
+		return r.delete(ctx, log, natGateway)
+	}
+	return r.reconcile(ctx, log, natGateway)
+}
+
+func (r *NatGatewayReconciler) delete(ctx context.Context, log logr.Logger, natGateway *networkingv1alpha1.NATGateway) (ctrl.Result, error) {
+	return ctrl.Result{}, nil
+}
+
+func hash(ip *networkingv1alpha1.NATGatewayDestinationIP) string {
+	return fmt.Sprintf("%s/%d/%d", ip.IP, ip.Port, ip.EndPort)
+}
+
+func (r *NatGatewayReconciler) assignPorts(ctx context.Context, log logr.Logger, natGateway *networkingv1alpha1.NATGateway, destinations map[types.UID]*networkingv1alpha1.NATGatewayDestination, slotsPerIP int, portsPerNetworkInterface int32) ([]networkingv1alpha1.NATGatewayDestination, int32, error) {
+	natGatewayRouting := &networkingv1alpha1.NATGatewayRouting{}
+	if err := r.Get(ctx, types.NamespacedName{Namespace: natGateway.Namespace, Name: natGateway.Name}, natGatewayRouting); client.IgnoreNotFound(err) != nil {
+		return nil, 0, fmt.Errorf("unable to get natgateway routing %s: %w", client.ObjectKeyFromObject(natGatewayRouting), err)
+	}
+
+	used := make(map[string]bool)
+	for _, dest := range natGatewayRouting.Destinations {
```

Comment:
This part actually reuses the `old` ip/port combination if it is still needed in order to have no change. If a target is not present anymore it will directly removed